x-pack/libbeat/reader/etw/session.go

```diff
@@ -0,0 +1,265 @@
+// to its corresponding uint8 constant value
+func getTraceLevel(level string) uint8 {
+	switch level {
+	case "critical":
+		return TRACE_LEVEL_CRITICAL
+	case "error":
+		return TRACE_LEVEL_ERROR
+	case "warning":
+		return TRACE_LEVEL_WARNING
+	case "information":
+		return TRACE_LEVEL_INFORMATION
+	case "verbose":
+		return TRACE_LEVEL_VERBOSE
+	default:
+		return TRACE_LEVEL_INFORMATION
+	}
+}
+
+// newSessionProperties initializes and returns a pointer to EventTraceProperties
+// with the necessary settings for starting an ETW session.
+// See https://learn.microsoft.com/en-us/windows/win32/api/evntrace/ns-evntrace-event_trace_properties
+func newSessionProperties(sessionName string) *EventTraceProperties {
+	// Calculate buffer size for session properties.
+	sessionNameSize := (len(sessionName) + 1) * 2
+	bufSize := sessionNameSize + int(unsafe.Sizeof(EventTraceProperties{}))
+
+	// Allocate buffer and cast to EventTraceProperties.
+	propertiesBuf := make([]byte, bufSize)
+	sessionProperties := (*EventTraceProperties)(unsafe.Pointer(&propertiesBuf[0]))
+
+	// Initialize mandatory fields of the EventTraceProperties struct.
+	// Filled based on https://learn.microsoft.com/en-us/windows/win32/etw/wnode-header
+	sessionProperties.Wnode.BufferSize = uint32(bufSize)
+	sessionProperties.Wnode.Guid = GUID{} // GUID not required for non-private/kernel sessions
+	// ClientContext is used for timestamp resolution
+	// Not used unless adding PROCESS_TRACE_MODE_RAW_TIMESTAMP flag to EVENT_TRACE_LOGFILE struct
+	// See https://learn.microsoft.com/en-us/windows/win32/etw/wnode-header
+	sessionProperties.Wnode.ClientContext = 1
+	sessionProperties.Wnode.Flags = WNODE_FLAG_TRACED_GUID
+	// Set logging mode to real-time
+	// See https://learn.microsoft.com/en-us/windows/win32/etw/logging-mode-constants
+	sessionProperties.LogFileMode = EVENT_TRACE_REAL_TIME_MODE
+	sessionProperties.LogFileNameOffset = 0                                            // Can be specified to log to a file as well as to a real-time session
+	sessionProperties.BufferSize = 64                                                  // Default buffer size, can be configurable
+	sessionProperties.LoggerNameOffset = uint32(unsafe.Sizeof(EventTraceProperties{})) // Offset to the logger name
+
+	return sessionProperties
+}
+
+// NewSession initializes and returns a new ETW Session based on the provided configuration.
+func NewSession(conf Config) (Session, error) {
+	var session Session
+	var err error
+
+	// Assign ETW Windows API functions
+	session.startTrace = _StartTrace
+	session.controlTrace = _ControlTrace
+	session.enableTrace = _EnableTraceEx2
+	session.openTrace = _OpenTrace
+	session.processTrace = _ProcessTrace
+	session.closeTrace = _CloseTrace
+
+	session.Name = setSessionName(conf)
+	session.Realtime = true
+
+	// Set default callbacks if not already specified.
+	if session.Callback == 0 {
+		session.Callback = syscall.NewCallback(DefaultCallback)
```

Comment:
Why should this package provide a default callback function? IIUC this is how users of the package will receive the data, and so it seems like it should be a hard requirement for the user to provide a non-nil callback otherwise it's an error.